crates/bevy_remote/src/lib.rs

```diff
@@ -0,0 +1,849 @@
+pub struct BrpRequest {
+    /// This field is mandatory and must be set to `"2.0"` for the request to be accepted.
+    pub jsonrpc: String,
+
+    /// The action to be performed.
+    pub method: String,
+
+    /// Arbitrary data that will be returned verbatim to the client as part of
+    /// the response.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub id: Option<Value>,
+
+    /// The parameters, specific to each method.
+    ///
+    /// These are passed as the first argument to the method handler.
+    /// Sometimes params can be omitted.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub params: Option<Value>,
+}
+
+/// A response according to BRP.
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct BrpResponse {
+    /// This field is mandatory and must be set to `"2.0"`.
+    pub jsonrpc: &'static str,
+
+    /// The id of the original request.
+    pub id: Option<Value>,
+
+    /// The actual response payload.
+    #[serde(flatten)]
+    pub payload: BrpPayload,
+}
+
+impl BrpResponse {
+    /// Generates a [`BrpResponse`] from an id and a `Result`.
+    #[must_use]
+    pub fn new(id: Option<Value>, result: BrpResult) -> Self {
+        Self {
+            jsonrpc: "2.0",
+            id,
+            payload: BrpPayload::from(result),
+        }
+    }
+}
+
+/// A result/error payload present in every response.
+#[derive(Debug, Serialize, Deserialize, Clone)]
+#[serde(rename_all = "snake_case")]
+pub enum BrpPayload {
+    /// `Ok` variant
+    Result(Value),
+    /// `Err` variant
+    Error(BrpError),
+}
+
+impl From<BrpResult> for BrpPayload {
+    fn from(value: BrpResult) -> Self {
+        match value {
+            Ok(v) => Self::Result(v),
+            Err(err) => Self::Error(err),
+        }
+    }
+}
+
+/// An error a request might return.
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct BrpError {
+    /// Defines the general type of the error.
+    pub code: i16,
+    /// Short, human-readable description of the error.
+    pub message: String,
+    /// Optional additional error data.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub data: Option<Value>,
+}
+
+impl BrpError {
+    /// Entity wasn't found.
+    #[must_use]
+    pub fn entity_not_found(entity: Entity) -> Self {
+        Self {
+            code: error_codes::ENTITY_NOT_FOUND,
+            message: format!("Entity {entity} not found"),
+            data: None,
+        }
+    }
+
+    /// Component wasn't found in an entity.
+    #[must_use]
+    pub fn component_not_present(component: &str, entity: Entity) -> Self {
+        Self {
+            code: error_codes::COMPONENT_NOT_PRESENT,
+            message: format!("Component `{component}` not present in Entity {entity}"),
+            data: None,
+        }
+    }
+
+    /// An arbitrary component error. Possibly related to reflection.
+    #[must_use]
+    pub fn component_error(error: anyhow::Error) -> Self {
```

Comment:
Why does this function take `anyhow::Error` while `BrpError::internal` takes `impl ToString`? Should we have both take `impl ToString` since the `anyhow::Error` is immediately converted to a string?